File: lab2/closest_pair.py
#!/usr/bin/env python3

import sys
import time
from math import hypot
from typing import List


Point = tuple[int, float, float]
Pair = tuple[Point, Point]


def read(line: str) -> Point:
    s = line.split()
    return int(s[0]), float(s[1]), float(s[2])


def dist(p1: Point, p2: Point) -> float:
    dx = p1[1] - p2[1]
    dy = p1[2] - p2[2]
    return hypot(dx, dy)


def brute_force(ps: List[Point]) -> tuple[float, Pair | None]:
    n = len(ps)
    d, pair = float("inf"), None
    for i in range(n - 1):
        for j in range(i + 1, n):
            if dist(ps[i], ps[j]) < d:
                d = dist(ps[i], ps[j])
                pair = ps[i], ps[j]

    return d, pair


def solve(xs: List[Point], ys: List[Point]) -> tuple[float, Pair | None]:
    n = len(xs)
    if n <= 1:
        return float("inf"), None
    if n == 2:
        return dist(*xs), tuple(xs)

    mid = n // 2
    xm = xs[mid - 1][1]

    ansl = solve(xs[:mid], [p for p in ys if p[1] <= xm])
    ansr = solve(xs[mid:], [p for p in ys if p[1] > xm])
    d, pair = min(ansl, ansr)

    ym = [p for p in ys if abs(p[1] - xm) < d]
    m = len(ym)
    for i in range(m - 1):
        for j in range(i + 1, min(i + 8, m)):
            if dist(ym[i], ym[j]) < d:
                d = dist(ym[i], ym[j])
                pair = ym[i], ym[j]

    return d, pair


def closest_pair(ps: List[Point]) -> tuple[float, Pair | None]:
    xs = sorted(ps, key=lambda p: p[1])
    ys = sorted(ps, key=lambda p: p[2])

    return solve(xs, ys)


if __name__ == "__main__":
    with open("data.txt") as f:
        data = [read(line) for line in f.readlines()]

        start = time.time()

        if len(sys.argv) < 2:
            d, pair = closest_pair(data)
        elif len(sys.argv) == 2 and sys.argv[1] == "brute_force":
            d, pair = brute_force(data)
        else:
            print("Usage: ./closest_pair.py [brute_force]")
            sys.exit(1)

        end = time.time()

        print(f"Algorithm finishes in {end - start}s.")
        print("Solution:")

        assert pair is not None
        print(*pair[0])
        print(*pair[1])
        print(f"d = {d}")

File: lab2/closest_pair.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <chrono>
#include <span>
#include <ranges>

struct Point {
    int id;
    float x;
    float y;

    Point()
        : id(-1), x(0.0f), y(0.0f) {}

    Point(int id, float x, float y)
        : id(id), x(x), y(y) {}
};

float dist(const Point& p1, const Point& p2)
{
    float dx = p1.x - p2.x;
    float dy = p1.y - p2.y;
    return std::hypot(dx, dy);
}

static std::pair<float, std::pair<Point, Point>> brute_force(std::span<Point> points)
{
    float d = std::numeric_limits<float>::infinity();
    size_t p1, p2;
    for (size_t i = 0; i < points.size() - 1; ++i) {
        for (size_t j = i + 1; j < points.size(); ++j) {
            if (dist(points[i], points[j]) < d) {
                d = dist(points[i], points[j]);
                p1 = i;
                p2 = j;
            }
        }
    }
    return {d, {points[p1], points[p2]}};
}

static std::pair<float, std::pair<Point, Point>> solve(std::span<Point> points)
{
    namespace ranges = std::ranges;

    size_t n = points.size();
    if (n < 32) {
        ranges::sort(points, {}, &Point::y);
        return brute_force(points);
    }

    const auto ansl = solve(points.first(n / 2));
    const auto ansr = solve(points.subspan(n / 2));
    auto [d, pair] = ansl.first < ansr.first ? ansl : ansr;

    const auto middle = ranges::next(points.begin(), n / 2);
    ranges::inplace_merge(points, middle, {}, &Point::y);

    const float middle_x = (middle - 1)->x;
    std::vector<Point> middle_points;
    ranges::copy_if(
        points,
        std::back_inserter(middle_points),
        [middle_x, d](float x) { return std::abs(x - middle_x) < d; },
        &Point::x);

    size_t m = middle_points.size();
    for (size_t i = 0; i < m - 1; ++i) {
        for (size_t j = i + 1; j < std::min(i + 8, m); ++j) {
            if (dist(points[i], points[j]) < d) {
                d = dist(points[i], points[j]);
                pair = {points[i], points[j]};
            }
        }
    }
    return {d, pair};
}

std::pair<float, std::pair<Point, Point>> closest_pair(std::span<Point> points)
{
    std::ranges::sort(points, {}, &Point::x);
    return solve(points);
}

int main()
{
    std::ifstream fin("data.txt");
    if (!fin) {
        std::cerr << "Cannot load data" << std::endl;
        return 1;
    }

    int id;
    float x, y;
    std::vector<Point> points;

    while (fin >> id >> x >> y) {
        points.emplace_back(id, x, y);
    }

    auto start = std::chrono::high_resolution_clock::now();

    auto [d, pair] = closest_pair(points);

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end - start;
    std::cout << "Algorithm finishes in " << diff << '\n';

    std::cout << "Solution:\n";
    std::cout << pair.first.id << " " << pair.first.x << " " << pair.first.y << '\n';
    std::cout << pair.second.id << " " << pair.second.x << " " << pair.second.y << '\n';
    std::cout << "d = " << d << '\n';
    return 0;
}
